### ISSUE TITLE
CFE doesn't insert type casts for loop variables and if conditions in control flow collections.

### ISSUE THREAD

**Issue by Markzipan** [dart-lang/sdk#36443]
The following snippet:
```
import 'expect.dart';

void main() {
  dynamic nonInt = "string";
  for (int i = nonInt; false; ) {}
  var x = <int>[for (int i = nonInt; false;) 1];
}
```
generates 
```
  static method main() → void {
    dynamic nonInt = "string";
    for (core::int i = nonInt as{TypeError} core::int; false; ) {
    }
    core::List<core::int> x = block {
      final core::List<core::int> #t1 = <core::int>[];
      for (core::int i = nonInt; false; )
        #t1.{core::List::add}(1);
    } =>#t1;
  }
```

Which is missing the type check.

Similarly for if expressions:
```
import 'expect.dart';

void main() {
  dynamic nonBool = 3;
  if (nonBool) {}
  Expect.throwsTypeError(() => <int>[if (nonBool) 1]);
}
```
which generates:
```
  static method main() → void {
    dynamic nonBool = 3;
    if(nonBool as{TypeError} core::bool) {
    }
    exp::Expect::throwsTypeError(() → core::List<core::int> => block {
      final core::List<core::int> #t1 = <core::int>[];
      if(nonBool)
        #t1.{core::List::add}(1);
    } =>#t1);
  }
```

**Comment by kmillikin** [comment]
The issue with for loops was fixed by 45bd79c119a5d562a492719df84b89448409eb4f.  The issue with if elements will be fixed by [CL 98484](https://dart-review.googlesource.com/c/sdk/+/98484).

**Comment by stefantsov** [comment]
As Kevin mentioned, the first part of the issue is fixed by https://github.com/dart-lang/sdk/commit/45bd79c119a5d562a492719df84b89448409eb4f.  The second part is fixed by e1e5bb59c9bbf152db8215bc3094f243fb650f9e that has just landed.
